aesd-char-driver/main.c

```diff
@@ -0,0 +1,307 @@
+	//TODO: return value?
+    }
+    
+    //check if data is \n terminated
+    partial_data = is_data_partial(buffptr, count);
+    if (partial_data == 0) {
+
+	// if there is data in partial buffer, handle that data
+	if (dev->part_mem_ptr != NULL) {
+            buffptr = handle_partial_data(dev->part_mem_ptr, 
+			    dev->part_buf_size, buffptr, (dev->part_buf_size+count));
+	    total_buf_size = dev->part_buf_size+count;
+	    //reset partial buffer info
+	    dev->part_mem_ptr = NULL;
+	    dev->part_buf_size = 0;
+	}
+
+	PDEBUG(" aesd_write: FINAL BUFF contains %s and count is %d\r\n", buffptr, count);
+	
+	wr_entry.buffptr = buffptr;
+	wr_entry.size = (total_buf_size != 0) ? total_buf_size:count;//handle partial data if any
+	
+	//take lock before writting the data into circular buffer
+	if (mutex_lock_interruptible(&dev->buf_lock)) {
+	    kfree(buffptr);
+            return 0;//-ERESTARTSYS;
+	}
+
+	rm_buffptr = aesd_circular_buffer_add_entry(&dev->buffer, &wr_entry);
+	mutex_unlock(&dev->buf_lock);
+
+        if (rm_buffptr != NULL) { //overwritten entry, free it
+	   kfree(rm_buffptr);
+	}
+
+	retval = count;
+    } else { //partial data
+	dev->part_mem_ptr = handle_partial_data(dev->part_mem_ptr,           
+                 dev->part_buf_size, buffptr, (dev->part_buf_size+count));
+	PDEBUG(" aesd_write: CURR BUFF contains %s\r\n", dev->part_mem_ptr);
+	dev->part_buf_size += count;
+	retval = count;
+    }
+
+    /**
+     * TODO: handle write
+     */
+    return retval;
+}
+struct file_operations aesd_fops = {
+    .owner =    THIS_MODULE,
+    .read =     aesd_read,
+    .write =    aesd_write,
+    .open =     aesd_open,
+    .release =  aesd_release,
+};
+
+static int aesd_setup_cdev(struct aesd_dev *dev)
+{
+    int err, devno = MKDEV(aesd_major, aesd_minor);
+
+    cdev_init(&dev->cdev, &aesd_fops);
+    dev->cdev.owner = THIS_MODULE;
+    dev->cdev.ops = &aesd_fops;
+    err = cdev_add (&dev->cdev, devno, 1);
+    if (err) {
+        printk(KERN_ERR "Error %d adding aesd cdev", err);
+    }
+    return err;
+}
+
+
+
+int aesd_init_module(void)
+{
+    dev_t dev = 0;
+    int result;
+    result = alloc_chrdev_region(&dev, aesd_minor, 1,
+            "aesdchar");
+    aesd_major = MAJOR(dev);
+    if (result < 0) {
+        printk(KERN_WARNING "Can't get major %d\n", aesd_major);
+        return result;
+    }
+    memset(&aesd_device,0,sizeof(struct aesd_dev));
+
+    /**
+     * TODO: initialize the AESD specific portion of the device
+     */
+
+    mutex_init(&aesd_device.buf_lock);
+    result = aesd_setup_cdev(&aesd_device);
+
+    if( result ) {
+        unregister_chrdev_region(dev, 1);
+    }
+    return result;
+
+}
+
+void aesd_cleanup_module(void)
+{
+    dev_t devno = MKDEV(aesd_major, aesd_minor);
+
+    cdev_del(&aesd_device.cdev);
+
+    /**
+     * TODO: cleanup AESD specific poritions here as necessary
+     */
+
+    // take the lock before cleaning the buffers(maybe not needed here)
```

Comment:
Good thought, shouldn't be needed though... since the kernel shouldn't let you run your cleanup until there are no active references to your read/write functions.